In [23]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [24]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [25]:
# Separate the data into labels and features

# Separate the y variable, the labels
target_df = lending_data_df["loan_status"]

# Separate the X variable, the features
features_data_df = lending_data_df.drop(columns = ["loan_status"])

#features_data_df.head()

In [26]:
# Review the y variable Series
target_df.head()


0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [27]:
# Review the X variable DataFrame
features_data_df.head()


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [28]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function

X_train, X_test, y_train, y_test = train_test_split(features_data_df,target_df,random_state = 1)



---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [29]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [30]:
# Make a prediction using the testing data
testing_predictions = logistic_regression_model.predict(X_test)
testing_predictions

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [31]:
# Generate a confusion matrix for the model
test_matrix = confusion_matrix(y_test, testing_predictions)
test_matrix

array([[18655,   110],
       [   36,   583]], dtype=int64)

In [32]:
# Print the classification report for the model
testing_classification_report = classification_report(y_test, testing_predictions)

print(testing_classification_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.94      0.89       619

    accuracy                           0.99     19384
   macro avg       0.92      0.97      0.94     19384
weighted avg       0.99      0.99      0.99     19384



In [33]:
lending_data_df["loan_status"].value_counts()


loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model seems to predict the 0 (healthy loan) label very well, with a precision of 1.00, recall of 0.99, and an f1-score of 1.00. This indicates that when the model predicts a loan as healthy, it is correct 100% of the time, and it correctly identifies 99% of all healthy loans.

For the 1 (high-risk loan) label, the precision is 0.84, recall is 0.94, and f1-score is 0.89. This suggests that while the model is relatively good at identifying high-risk loans, there is still room for improvement. It correctly identifies 94% of all high-risk loans but misclassifies 16% of the loans it predicts as high risk.

---